File: dot_local/share/codex/skills/sem/references/sem/crates/sem-core/tests/fixtures/scope_test/ruby/service.rb
require_relative 'models'
require_relative 'database'

def create_dog(name)
  dog = Dog.new(name)
  raise ArgumentError, "invalid dog" unless dog.validate
  conn = get_connection
  conn.execute("INSERT INTO dogs VALUES (?)")
  conn.commit
  dog
end

def create_cat(name)
  cat = Cat.new(name)
  raise ArgumentError, "invalid cat" unless cat.validate
  conn = get_connection
  conn.execute("INSERT INTO cats VALUES (?)")
  conn.commit
  cat
end

def transfer_animal(animal, shelter)
  txn = Transaction.new(get_connection)
  txn.execute("UPDATE animals SET shelter_id = ?")
  shelter.add(animal)
  txn.commit
end

def list_animals
  conn = get_connection
  conn.execute("SELECT * FROM animals")
end

File: dot_local/share/codex/skills/sem/references/sem/crates/sem-core/tests/fixtures/scope_test/cpp/service.cpp
#include "models.hpp"
#include "database.hpp"

Dog createDog(std::string name) {
    Dog dog(name);
    if (!dog.validate()) {
        throw std::invalid_argument("invalid dog");
    }
    Connection* conn = getConnection();
    conn->execute("INSERT INTO dogs VALUES (?)");
    conn->commit();
    return dog;
}

Cat createCat(std::string name) {
    Cat cat(name);
    if (!cat.validate()) {
        throw std::invalid_argument("invalid cat");
    }
    Connection* conn = getConnection();
    conn->execute("INSERT INTO cats VALUES (?)");
    conn->commit();
    return cat;
}

void transferAnimal(void* animal, Shelter* shelter) {
    Transaction txn(getConnection());
    txn.execute("UPDATE animals SET shelter_id = ?");
    shelter->add(animal);
    txn.commit();
}

void* listAnimals() {
    Connection* conn = getConnection();
    return conn->execute("SELECT * FROM animals");
}
